qiskit/primitives/gradient/param_shift_sampler_gradient.py

```diff
@@ -0,0 +1,131 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Gradient of probabilities with parameter shift
+"""
+
+from __future__ import annotations
+
+from collections import Counter, defaultdict
+from dataclasses import dataclass
+from typing import Sequence, Type
+
+import numpy as np
+
+from qiskit import transpile
+from qiskit.circuit import Parameter, ParameterExpression, QuantumCircuit
+from qiskit.result import QuasiDistribution
+
+from ..base_sampler import BaseSampler
+from ..sampler_result import SamplerResult
+
+
+@dataclass
+class SubSampler:
+    coeff: float | ParameterExpression
+    circuit: QuantumCircuit
+    index: int
+
+
+class ParamShiftSamplerGradient:
+    """Parameter shift estimator gradient"""
+
+    SUPPORTED_GATES = ["x", "y", "z", "h", "rx", "ry", "rz", "p", "cx", "cy", "cz"]
+
+    def __init__(self, sampler: Type[BaseSampler], circuit: QuantumCircuit):
+        self._circuit = circuit
+        self._grad = self._preprocessing()
+        circuits = [self._circuit]
+        for param, lst in self._grad.items():
+            for arg in lst:
+                circuits.append(arg.circuit)
+        self._sampler = sampler(circuits=circuits)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *exc_info):
+        pass
+
+    @classmethod
+    def _gradient_circuits(cls, circuit: QuantumCircuit):
+        circuit2 = transpile(circuit, basis_gates=cls.SUPPORTED_GATES, optimization_level=0)
+        ret = defaultdict(list)
+        for inst in circuit2.data:
+            if inst[0].is_parameterized():
+                param = inst[0].params[0]
+                for p in param.parameters:
+                    # TODO: Need to wait for an appropriate way to update parameters of
+                    #   a particular instruction.
+                    #   See https://github.com/Qiskit/qiskit-terra/issues/7894
+                    inst[0].params[0] = param + np.pi / 2
+                    ret[p].append((circuit2.copy(), param.gradient(p) / 2))
```

Comment:
I'm not sure copying all the circuits is the best way to go here, since this is quite a costly operation. Couldn't we just execute the same circuit with different parameters and leverage that the primitives cache the transpiled circuits?

Reply:
Yeah, we are not focusing on the performance at this moment, but, copying circuits would be a performance issue. If a parameter is used in multiple gates, we need to modify the gates independently. I'm thinking of an idea to assign intermediate parameters to such gates. For example, assume that there are `Ry(x, qr[0])` and `Rx(x * 2, qr[1])`, we replace them with `Ry(t[0], qr[0])` and `Rx(t[1], qr[1])` with `t = [x, x * 2]`. Then we can update each gate independently by setting `t` values by ourselves. This is inspired from [SSA](https://en.wikipedia.org/wiki/Static_single_assignment_form)